dartagnan/src/main/antlr4/LitmusC.g4

```diff
@@ -239,6 +261,28 @@ c11Mo returns [String mo]
     |   MoSeqCst    {$mo = C11.MO_SC;}
     ;
 
+openCLSpace returns [String space]
+    :   OpenCLGlobalSpace  {$space = OpenCL.GLOBAL_SPACE;}
+    |   OpenCLLocalSpace   {$space = OpenCL.LOCAL_SPACE;}
+    ;
+
+openCLScope returns [String scope]
+    :   OpenCLMemoryScopeWI    {$scope = OpenCL.WORK_ITEM;}
+    |   OpenCLMemoryScopeWG    {$scope = OpenCL.WORK_GROUP;}
+    |   OpenCLMemoryScopeDEV   {$scope = OpenCL.DEVICE;}
+    |   OpenCLMemoryScopeALL   {$scope = OpenCL.ALL;}
+    ;
+
+openCLFenceFlags
+    :   openCLFenceFlag Bar openCLFenceFlag
+    |   openCLFenceFlag
+    ;
+
+openCLFenceFlag returns [String flag]
+    :   OpenCLFenceFlagGL    {$flag = OpenCL.GLOBAL_SPACE;}
+    |   OpenCLFenceFlagLC    {$flag = OpenCL.LOCAL_SPACE;}
+    ;
+
```

Comment:
The rules refers to `flag`, but the return value to addresses. Would it make sense to merge this somehow, i.e. use `openCLSpace ` wherever we are using `openCLFenceFlag `?

Reply:
Do you mean changing the returning `flag` to `space `to be consistent with `openCLSpace`? Or using `openCLSpace` to replace `openCLFenceFlag`? 
`openCLSpace` is `global`/`local` while `openCLFenceFlag` is `CLK_GLOBAL_MEM_FENCE`/`CLK_LOCAL_MEM_FENCE`, is there a good way to do that?

Reply:
> `openCLFenceFlag` is `CLK_GLOBAL_MEM_FENCE`/`CLK_LOCAL_MEM_FENCE`, is there a good way to do that?

Shouldn't then this be 

```
openCLFenceFlag returns [String flag]
    :   OpenCLFenceFlagGL    {$flag = CLK_GLOBAL_MEM_FENCE;}
    |   OpenCLFenceFlagLC    {$flag = CLK_LOCAL_MEM_FENCE;}
    ;
```
???

Reply:
Why do we return flags like that? I meant the token of `OpenCLFenceFlagGL` is `CLK_GLOBAL_MEM_FENCE`, and `OpenCLFenceFlagLC` is `CLK_LOCAL_MEM_FENCE`.

Reply:
Good point. I guess this is an abuse of code coming from "similar" situations like c11 memory orders. However, it makes sense for c11 because there is a difference between the tokens and the tags, e.g. `memory_order_acquire` and `ACQ`.

You should get rid of the flags and use the token (please also fix all other cases where we have the same issue)

Reply:
I think the process is similar for OpenCL? 
For token `global`(from memory event) and `CLK_GLOBAL_MEM_FENCE`(from fence) we should add `GLOBAL` tag;
for token `local` and `CLK_LOCAL_MEM_FENCE` we should add `LOCAL` tag.

Reply:
I don't understand what you mean

Reply:
I meant for example we got this one fence:
```atomic_work_item_fence(CLK_GLOBAL_MEM_FENCE,memory_order_release,memory_scope_device);```
Then we should add `GLOBAL` tag to it in our visitor, and `CLK_GLOBAL_MEM_FENCE` is not a tag.

Reply:
Now I am confused. Your current visitor seems not to be adding `GLOBAL` (which should claim should be added) but it is actually adding `CLK_GLOBAL_MEM_FENCE` which should not (the situation is even worse, because if the instruction uses `CLK_GLOBAL_MEM_FENCE  | CLK_LOCAL_MEM_FENCE `, it adds both thins as a single string/tag)

Reply:
It is using the `OpenCLFenceFlagContext().flag` in the current visitor, which I think should be `GLOBAL`/`LOCAL`. Why do you say it adds both as a single string when fence has both spaces? I think they were added one by one in iteration.

Reply:
> It is using the `OpenCLFenceFlagContext().flag` in the current visitor, which I think should be `GLOBAL`/`LOCAL`. 

Ok, I now understand why both `openCLSpace` and `openCLFenceFlag` make use of `OpenCL.GLOBAL_SPACE/OpenCL.LOCAL_SPACE`. However, for the first one we can simply used the token and get rid of the `{$space = OpenCL.GLOBAL_SPACE;}` part, right?

> Why do you say it adds both as a single string when fence has both spaces? I think they were added one by one in iteration.

I misread the visitor, I thought `openCLFenceFlag()` was being used in `fence.addTags()` instead of `flagCtx.flag`. Sorry about the confusion